[MRG] update `Index` protocol tests to include tests for `peek` and `consume`

ctb opened the pull request:
Test, document, rationalize, and generalize the `CounterGather` API used for 'gather' optimization.

The main work of this PR is to move tests specific to the `CounterGather` implementation in `src/sourmash/index/__init__.py` over to more generic protocol tests that test the API rather than the implementation. As part of this, the PR implements an additional `CounterGather` compatible class, based on `LinearIndex`.

The ultimate goals are to -
* nail down the `CounterGather` API a bit more;
* provide options for alternative implementations that provide specific optimizations;

This PR does the following:
- changes the `peek` and `add` methods to use named parameters with a `*`;
- moves the `CounterGather` tests from `test_index.py` into `test_index_protocol`;
- builds a `CounterGather`-compatible class wrapper for `LinearIndex`, for testing purposes;
- moves some common threshold calculation code from the `CounterGather` class over to `search.py`;
- adds a new test, `test_counter_gather_exact_match`;

Fixes https://github.com/sourmash-bio/sourmash/issues/1960

## Notes for review

The tests that moved from `test_index.py` to `test_index_protocol.py` are unchanged functionally; it's not well represented in the diff, tho, because they moved files.

TODO:
- [x] experiment with adding multiple identical matches, what happens?

ctb committed:
move most CounterGather tests over to index protocol tests

ctb committed:
add LinearIndex wrapper

ctb committed:
getting closer

ctb committed:
fix a bunch of the tests

ctb committed:
Merge branch 'latest' of https://github.com/sourmash-bio/sourmash into refactor/counter_gather_tests

ctb committed:
Merge branch 'latest' of https://github.com/sourmash-bio/sourmash into refactor/counter_gather_tests

ctb committed:
fix call to 'peek'

ctb committed:
adjust 'counter.add' call signature

ctb committed:
add CounterGather_LCA

ctb committed:
move CounterGather.calc_threshold into search.py

ctb committed:
minor refactoring

ctb committed:
resolve downsampling for linear index wrapper

ctb committed:
fix downsampling for LCA-based CounterGather

ctb committed:
fix location foo

ctb committed:
fix remaining test

ctb committed:
minor cleanup

ctb committed:
add doc

ctb committed:
test multiple identical matches

ctb committed:
adjust LinearIndex implementation to skip identical matches

ctb committed:
cleanup protocol tests

ctb committed:
revert LCA_Database fix

ctb committed:
Merge branch 'latest' into refactor/counter_gather_tests

ctb committed:
cleanup

ctb committed:
Merge branch 'refactor/counter_gather_tests' of https://github.com/sourmash-bio/sourmash into refactor/counter_gather_tests

ctb committed:
Merge branch 'latest' into refactor/counter_gather_tests

ctb committed:
Merge branch 'latest' into refactor/counter_gather_tests

ctb commented:
bump @mr-eyes

ctb commented:
whoops, thought that I'd already asked for review 😊 

@mr-eyes could you review this? it should all be functionality-neutral API changes / test refactoring.

bluegenes commented on src/sourmash/search.py:
```diff
@@ -11,6 +11,25 @@
+    if threshold_bp:
+        # if we have a threshold_bp of N, then that amounts to N/scaled
+        # hashes:
+        n_threshold_hashes = float(threshold_bp) / scaled
```
is `threshold_bp` constrained to a positive value? I see `0` is handled and assume a neg thresh wouldn't be possible, but wanted to ask

ctb replied:
good catch :)

it's not constrained in this PR, which is mostly about adding tests so that I can refactor the code - but as [you noted](https://github.com/sourmash-bio/sourmash/pull/2116/files#r922571983) it's handled over in #2116!

bluegenes reviewed (approved):
just had the one q, otherwise looks good, i think!

mr-eyes commented on src/sourmash/index/__init__.py:
```diff
@@ -701,31 +706,42 @@ def select(self, **kwargs):
+    other implementations are possible ;).
     """
     def __init__(self, query_mh):
+        "Constructor - takes a query FracMinHash."
```
Is it limited to accepting FracMinHash? If so, would you please point me to where I can find why it's limited to FracMinHashes?

Also, if it is limited to FracMinHashes, do you think we should add an if condition?

ctb replied:
Yep - it's limited to estimators that support containment operations, and num MinHashes do not.

The if statement a few lines below does the correct check:
```python
        if not query_mh.scaled:
            raise ValueError('gather requires scaled signatures')
```
which is tested by `test_counter_gather_num_query`.

In addition, (spoiler alert!) https://github.com/sourmash-bio/sourmash/pull/2116 refactors the class to take a `SourmashSignature` instead of a `MinHash`, which (in some possible future) opens it up to other kinds of estimators.

mr-eyes commented on src/sourmash/search.py:
```diff
@@ -11,6 +11,25 @@
 from .sketchcomparison import FracMinHashComparison, NumMinHashComparison
 
 
+def calc_threshold_from_bp(threshold_bp, scaled, query_size):
```
Is it possible for the `threshold_bp` to be larger than the `query_size`? Throwing an exception would be helpful here.

ctb replied:
yes, good catch! this is addressed over in #2116, where I refactor some other code to use this same function:

https://github.com/sourmash-bio/sourmash/pull/2116/files#diff-51443303a0e6c38023d8977d891df4ae18dc06a2431d29338fa8648a042132bfR33

mr-eyes commented:
> @mr-eyes could you review this? it should all be functionality-neutral API changes / test refactoring.

Sorry, it took some time.
Just some minor comments/questions. Otherwise, looks good to me.

ctb commented:
thanks to both of you - I was definitely fixated on expanding the tests on this one, and not doing much with the core code, so it's nice that I paired it with #2116 which addresses your concerns 😆
